cmd/bblfshd/main.go

```
@@ -0,0 +1,158 @@
+	cmd.Parse(os.Args[1:])
+
+	runtime.Bootstrap()
+	logger = buildLogger()
+}
+
+func main() {
+	logger.Infof("bblfshd version: %s (build: %s)", version, build)
+
+	r := buildRuntime()
+	s := daemon.NewDaemon(version, r)
+	s.Options = buildGRPCOptions()
+	s.Overrides = buildOverrides()
+	s.Logger = logger
+
+	l, err := net.Listen(*network, *address)
+	if err != nil {
+		logger.Errorf("error creating listener: %s", err)
+		os.Exit(1)
+	}
+
+	logger.Infof("server listening in %s (%s)", *address, *network)
+	handleGracefullyShutdown(l, s)
+
+	if err = s.Serve(l); err != nil {
+		logger.Errorf("error starting server: %s", err)
+		os.Exit(1)
+	}
+}
+
+func buildLogger() server.Logger {
+	logger, err := server.LoggerFactory{
+		Level:  *log.level,
+		Format: *log.format,
+		Fields: *log.fields,
+	}.New()
+
+	if err != nil {
+		logrus.Errorf("invalid logger configuration: %s", err)
+		os.Exit(1)
+	}
+
+	return logger
+}
+
+func buildRuntime() *runtime.Runtime {
+	logger.Infof("initializing runtime at %s", *storage)
+
+	r := runtime.NewRuntime(*storage)
+	if err := r.Init(); err != nil {
+		logger.Errorf("error initializing runtime: %s", err)
+		os.Exit(1)
+	}
+
+	return r
+}
+
+func buildGRPCOptions() []grpc.ServerOption {
+	size := *maxMessageSize
+	if size >= 2048 {
+		// Setting the hard limit of message size to less than 2GB since
+		// it may overflow an int value, and it should be big enough
+		logger.Errorf("max-message-size too big (limit is 2047MB): %d", size)
+		os.Exit(1)
+	}
+
+	size = size * 1024 * 1024
+
+	return []grpc.ServerOption{
+		grpc.MaxRecvMsgSize(size),
+		grpc.MaxSendMsgSize(size),
+	}
+}
+
+func buildOverrides() map[string]string {
+	overrides := make(map[string]string)
+	for _, img := range strings.Split(os.Getenv("BBLFSH_DRIVER_IMAGES"), ";") {
+		if img = strings.TrimSpace(img); img == "" {
+			continue
+		}
+
+		fields := strings.Split(img, "=")
+		if len(fields) != 2 {
+			logger.Errorf("invalid image driver format %s", img)
+			os.Exit(1)
```

Having so many exit points makes code hard to reuse/test. It may be better to just return the error and let the caller handle it properly.

---

Witch caller, is a main

---

`main` function is the caller. I guess this is a matter of taste, since it's a direct call from main.

---

I just wanted to make the main function simple and avoid error handling there...